TypeErrors for @scullyio/scully-plugin-base-href-rewrite

## 🐞 Bug report

### Description

Thank you for the amazing project.

- Get type errors following the example
   - https://github.com/scullyio/scully/blame/main/libs/plugins/base-href-rewrite/README.md#L20-L32

Given my scully.my-project.config.ts,

```ts
import { ScullyConfig, setPluginConfig } from '@scullyio/scully'
import { baseHrefRewrite } from '@scullyio/scully-plugin-base-href-rewrite'

const defaultPostRenderers = ['seoHrefOptimise', baseHrefRewrite]
setPluginConfig(baseHrefRewrite, {
  href: '/json-to-protobuf',
})

export const config: ScullyConfig = {
  defaultPostRenderers,
  projectRoot: './src',
  projectName: 'json-to-protobuf',
  outDir: './dist/static',
  routes: {},
}
```

```shell
scully.json-to-protobuf.config.ts:5:17 - error TS2345: Argument of type 'unique symbol' is not assignable to parameter of type 'string'.

5 setPluginConfig(baseHrefRewrite, {
                  ~~~~~~~~~~~~~~~

scully.json-to-protobuf.config.ts:10:3 - error TS2322: Type '(string | symbol)[]' is not assignable to type 'string[]'.
  Type 'string | symbol' is not assignable to type 'string'.
    Type 'symbol' is not assignable to type 'string'.

10   defaultPostRenderers,
     ~~~~~~~~~~~~~~~~~~~~

  node_modules/@scullyio/scully/lib/utils/interfacesandenums.d.ts:18:5
    18     defaultPostRenderers?: string[];
           ~~~~~~~~~~~~~~~~~~~~
    The expected type comes from property 'defaultPostRenderers' which is declared here on type 'ScullyConfig'
```

#### 🔬 Minimal Reproduction

```shell
cd $(mktemp -d)
git clone git@github.com:kkweon/json-to-protobuf.git
cd json-to-protobuf
git checkout feat-scully-bug-report-plugin-base-href-rewrite
yarn
tsc scully.json-to-protobuf.config.ts --noEmit
```

#### 💻Your Environment

**Angular Version:**

<pre><code>
❯ ng --version

     _                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/


Angular CLI: 10.0.2
Node: 12.18.0
OS: darwin x64

Angular: 10.0.3
... animations, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, router
Ivy Workspace: Yes

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.1000.2
@angular-devkit/build-angular     0.1000.2
@angular-devkit/build-optimizer   0.1000.2
@angular-devkit/build-webpack     0.1000.2
@angular-devkit/core              10.0.2
@angular-devkit/schematics        10.0.2
@angular/cli                      10.0.2
@ngtools/webpack                  10.0.2
@schematics/angular               10.0.2
@schematics/update                0.1000.2
rxjs                              6.5.5
typescript                        3.9.6
webpack                           4.43.0
</code></pre>

**Scully Version:**

<pre><code>
    "@scullyio/init": "^0.0.29",
    "@scullyio/ng-lib": "latest",
    "@scullyio/scully": "latest",
    "@scullyio/scully-plugin-base-href-rewrite": "^0.0.2",
</code></pre>

## 🔥 Exception or Error

See above

--- a/libs/scully/src/lib/pluginManagement/index.ts
+++ b/libs/scully/src/lib/pluginManagement/index.ts
@@ -1,2 +1,7 @@
-export { registerPlugin, scullySystem } from './pluginRepository';
+export {
+  registerPlugin,
+  configValidator,
+  scullySystem,
+} from './pluginRepository';
+
 export { findPlugin } from './pluginConfig';

--- a/libs/scully/src/lib/pluginManagement/pluginConfig.ts
+++ b/libs/scully/src/lib/pluginManagement/pluginConfig.ts
@@ -17,7 +17,7 @@ export const routeConfigData = configData + 'Route_Config_Data__';
 export const resetConfig = configData + 'resetData__';
 
 export const setPluginConfig = (
-  name: string,
+  name: string | symbol,
   typeOrConfig: PluginTypes | Serializable,
   config?: Serializable
 ): void => {
@@ -48,17 +48,24 @@ export function fetchPlugins(name: string | symbol, type?: string): Function[] {
   return result;
 }
 
-export function findPlugin(name: string | symbol, type?: string): Function {
+export function findPlugin(
+  name: string | symbol,
+  type?: string,
+  errorOnNotfound = true
+): Function {
   const found = fetchPlugins(name, type);
   const displayName = typeof name === 'string' ? name : name.description;
   switch (found.length) {
     case 0:
-      logError(
-        `Plugin "${yellow(displayName)}" of type "${yellow(
-          type
-        )}" is not found, can not store config`
-      );
-      process.exit(15);
+      if (errorOnNotfound) {
+        logError(
+          `Plugin "${yellow(displayName)}" of type "${yellow(
+            type
+          )}" is not found, can not store config`
+        );
+        process.exit(15);
+      }
+      return undefined;
       break;
     case 1:
       const pl = found[0] as Function;
@@ -66,12 +73,15 @@ export function findPlugin(name: string | symbol, type?: string): Function {
         ? (pl[accessPluginDirectly] as Function)
         : pl;
     default:
-      logError(
-        `Plugin "${yellow(
-          displayName
-        )}" has multiple types, please specify type to be able to store config`
-      );
-      process.exit(15);
+      if (errorOnNotfound) {
+        logError(
+          `Plugin "${yellow(
+            displayName
+          )}" has multiple types, please specify type to be able to store config`
+        );
+        process.exit(15);
+      }
+      return undefined;
   }
 }
 

--- a/libs/scully/src/lib/renderPlugins/executePlugins.ts
+++ b/libs/scully/src/lib/renderPlugins/executePlugins.ts
@@ -1,13 +1,12 @@
 import {
-  plugins,
   registerPlugin,
   scullySystem,
 } from '../pluginManagement/pluginRepository';
+import { findPlugin } from '../pluginManagement/pluginConfig';
 import { HandledRoute } from '../routerPlugins/addOptionalRoutesPlugin';
 import { scullyConfig } from '../utils/config';
 import { logError, yellow } from '../utils/log';
 import { puppeteerRender } from './puppeteerRenderPlugin';
-import { findPlugin } from '../pluginManagement/pluginConfig';
 
 export const renderRoute = Symbol('renderRoute');
 
@@ -42,7 +41,7 @@ const executePluginsForRoute = async (route: HandledRoute) => {
   const InitialPromise = findPlugin(puppeteerRender)(route);
   return handlers.reduce(async (updatedHTML, plugin) => {
     const html = await updatedHTML;
-    const handler = plugins.render[plugin];
+    const handler = findPlugin(plugin, 'render', false);
     if (handler) {
       try {
         /** return result of plugin */

--- a/libs/scully/src/lib/routerPlugins/addOptionalRoutesPlugin.ts
+++ b/libs/scully/src/lib/routerPlugins/addOptionalRoutesPlugin.ts
@@ -1,5 +1,5 @@
 import { Serializable } from 'puppeteer';
-import { plugins } from '../pluginManagement/pluginRepository';
+import { findPlugin } from '../pluginManagement/pluginConfig';
 import { scullyConfig } from '../utils/config';
 import { RoutesTypes, RouteTypes } from '../utils/interfacesandenums';
 import { logError, logWarn, yellow } from '../utils/log';
@@ -12,11 +12,13 @@ export const addOptionalRoutes = async (
       const x = await result;
       const config = scullyConfig.routes[cur];
       if (config) {
-        const postRenderers: string[] = Array.isArray(config.postRenderers)
-          ? config.postRenderers
+        const postRenderers: (string | symbol)[] = Array.isArray(
+          config.postRenderers
+        )
+          ? (config.postRenderers as (string | symbol)[])
           : undefined;
         /** adding in the postrenderes. Note that the plugin might choose to overwrite the ones that come from the config */
-        const r = (await routePluginHandler(cur)).map(row =>
+        const r = (await routePluginHandler(cur)).map((row) =>
           postRenderers ? { postRenderers, ...row, config } : { ...row, config }
         );
         x.push(...r);
@@ -65,7 +67,7 @@ export interface HandledRoute {
     [key: string]: Serializable;
   };
   /** an array with render plugin names that will be executed */
-  postRenderers?: string[];
+  postRenderers?: (string | symbol)[];
   /** the path to the file for a content file */
   templateFile?: string;
   /** optional title, if data holds a title, that will be used instead */
@@ -90,12 +92,13 @@ async function routePluginHandler(route: string): Promise<HandledRoute[]> {
     logError(`No configuration for route "${yellow(route)}" found. Skipping`);
     return [{ route, type: RouteTypes.default }];
   }
-  if (plugins.router[conf.type]) {
-    const generatedRoutes = (await (plugins.router[conf.type](
+  const routerPlugin = findPlugin(conf.type, 'router', false);
+  if (routerPlugin) {
+    const generatedRoutes = (await (routerPlugin(
       route,
       conf
     ) as unknown)) as HandledRoute[];
-    generatedRoutes.forEach(handledRoute => {
+    generatedRoutes.forEach((handledRoute) => {
       if (!handledRoute.route.startsWith('/')) {
         logWarn(
           `The plugin '${

--- a/libs/scully/src/lib/utils/handlers/routeDiscovery.ts
+++ b/libs/scully/src/lib/utils/handlers/routeDiscovery.ts
@@ -35,6 +35,7 @@ export async function routeDiscovery(
     ).filter(
       (r) =>
         !r.route.endsWith('*') &&
+        /** use the routefilter to only include matches */
         (routeFilter === '' ||
           routeFilterRegexs.some((reg) => r.route.match(reg) !== null))
     );

--- a/libs/scully/src/lib/utils/interfacesandenums.ts
+++ b/libs/scully/src/lib/utils/interfacesandenums.ts
@@ -17,7 +17,7 @@ export interface ScullyConfig {
   /** the folder where the project sources resides, read from angular.json */
   sourceRoot?: string;
   /** Array with string ID's of the content-renderers that will be run on all routes */
-  defaultPostRenderers?: string[];
+  defaultPostRenderers?: (string | symbol)[];
   /** the root of the project (where angular.json lives) */
   homeFolder?: string;
   /** the destination of the Scully generated files */
@@ -70,12 +70,12 @@ export type RoutesTypes =
 
 export interface RouterTypeDefault {
   type: RouteTypes.default;
-  postRenderers?: string[];
+  postRenderers?: (string | symbol)[];
 }
 
 export type RouteTypeJson = {
   type: RouteTypes.json;
-  postRenderers?: string[];
+  postRenderers?: (string | symbol)[];
 } & {
   [paramName: string]: {
     url: string;
@@ -91,7 +91,7 @@ export interface HeadersObject {
 }
 
 export type RouteTypeContentFolder = {
-  postRenderers?: string[];
+  postRenderers?: (string | symbol)[];
   type: RouteTypes.contentFolder;
 } & {
   [paramName: string]: {
@@ -100,7 +100,7 @@ export type RouteTypeContentFolder = {
 };
 
 export type RouteTypeUnknown = {
-  postRenderers?: string[];
+  postRenderers?: (string | symbol)[];
   type: string;
 } & {
   [paramName: string]: any;

--- a/libs/scully/tsconfig.json
+++ b/libs/scully/tsconfig.json
@@ -1,5 +1,4 @@
 {
-  "extends": "../../tsconfig.json",
   "compileOnSave": false,
   "compilerOptions": {
     "baseUrl": "./scully",

--- a/scully.sample-blog.config.ts
+++ b/scully.sample-blog.config.ts
@@ -16,6 +16,7 @@ import './demos/plugins/voidPlugin';
 
 const FlashPrevention = getFlashPreventionPlugin();
 setPluginConfig('md', { enableSyntaxHighlighting: true });
+setPluginConfig(baseHrefRewrite, { href: '/' });
 
 const defaultPostRenderers = ['seoHrefOptimise'];
 

